go-client/client/client.go

```diff
@@ -0,0 +1,262 @@
+
+type fieldId struct {
+	appId     string
+	fieldName string
+}
+
+// Maintains a connection to a Deephaven server.
+// It can be used to run scripts, create new tables, execute queries, etc.
+// Check the various methods of Client to learn more.
+type Client struct {
+	grpcChannel *grpc.ClientConn
+
+	sessionStub
+	consoleStub
+	flightStub
+	tableStub
+	appStub
+	inputTableStub
+
+	nextTicket int32
+
+	tablesLock sync.Mutex
+	tables     map[fieldId]*TableHandle
+}
+
+// Starts a connection to a Deephaven server.
+// scriptLanguage can be either "python" or "groovy", and must match the language used on the server. Python is the default.
+//
+// The client should be closed using Close() after it is done being used.
+//
+// Note that the provided context is saved and used to send keepalive messages.
+func NewClient(ctx context.Context, host string, port string, scriptLanguage string) (*Client, error) {
+	grpcChannel, err := grpc.Dial(host+":"+port, grpc.WithTransportCredentials(insecure.NewCredentials()))
+	if err != nil {
+		return nil, err
+	}
+
+	client := &Client{grpcChannel: grpcChannel, tables: make(map[fieldId]*TableHandle)}
+
+	client.sessionStub, err = newSessionStub(ctx, client)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.tableStub, err = newTableStub(client)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.consoleStub, err = newConsoleStub(ctx, client, scriptLanguage)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.flightStub, err = newFlightStub(client, host, port)
+	if err != nil {
+		client.Close()
+		return nil, err
+	}
+
+	client.inputTableStub = newInputTableStub(client)
+
+	client.appStub = newAppStub(client)
+
+	return client, nil
+}
+
+// Checks if the client is closed, i.e. it can no longer perform operations on the server.
+func (client *Client) Closed() bool {
+	return client.grpcChannel == nil
+}
+
+type FetchOption int
+
+const (
+	FetchOnce      FetchOption = iota // Fetches the list of tables once and then returns.
+	FetchRepeating FetchOption = iota // Starts up a background thread to continually update the list of tables as changes occur.
+)
+
+// Fetches the list of tables from the server.
+// This allows the client to see the list of named global tables on the server,
+// and thus allows it to open them using OpenTables.
+// Tables created in scripts run by the current client are immediately visible and do not require a FetchTables call.
+func (client *Client) FetchTables(ctx context.Context, opt FetchOption) error {
+	if client.Closed() {
+		return ErrClosedClient
+	}
+
+	err := client.listFields(ctx, func(update *applicationpb2.FieldsChangeUpdate) {
+		client.handleFieldChanges(update)
+	})
+	if err != nil {
+		return err
+	}
+
+	if opt == FetchOnce {
+		client.appStub.cancelFunc()
+	}
+
+	return nil
+}
+
+// Returns a list of the (global) tables that can be opened with OpenTable.
+// This can be updated using FetchTables.
```

Comment:
1) Does `FetchRepeating` keep this list constantly up to date?
2) Does this return the values at the last time `FetchTables` was called (assuming `FetchOnce`)?
3) How else can this list get out of sync?